File: frontend/ZuhooCMS/src/app/modules/attendance/services/biometric-device.service.ts
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { ApiService, PagedResponse } from '../../../core/services/api.service';
import { BiometricDevice, BiometricDeviceRequest, BiometricDeviceStatus } from '../models/attendance.model';

@Injectable({ providedIn: 'root' })
export class BiometricDeviceService {
  private readonly endpoint = '/company/biometric/devices';
  constructor(private api: ApiService) {}

  create(payload: BiometricDeviceRequest): Observable<BiometricDevice> {
    return this.api.post<BiometricDevice>(this.endpoint, payload);
  }

  getById(id: number): Observable<BiometricDevice> {
    return this.api.get<BiometricDevice>(`${this.endpoint}/${id}`);
  }

  list(page = 0, size = 20): Observable<PagedResponse<BiometricDevice>> {
    return this.api.getPaged<BiometricDevice>(this.endpoint, page, size);
  }

  getByStatus(status: BiometricDeviceStatus, page = 0, size = 20): Observable<PagedResponse<BiometricDevice>> {
    return this.api.getPaged<BiometricDevice>(`${this.endpoint}/status/${status}`, page, size);
  }

  getOnline(): Observable<BiometricDevice[]> {
    return this.api.get<BiometricDevice[]>(`${this.endpoint}/online`);
  }

  update(id: number, payload: BiometricDeviceRequest): Observable<BiometricDevice> {
    return this.api.patch<BiometricDevice>(`${this.endpoint}/${id}`, payload);
  }

  updateStatus(id: number, status: BiometricDeviceStatus): Observable<void> {
    return this.api.patch<void>(`${this.endpoint}/${id}/status?status=${status}`, {});
  }

  recordSync(id: number): Observable<void> {
    return this.api.post<void>(`${this.endpoint}/${id}/sync`, {});
  }

  delete(id: number): Observable<BiometricDevice> {
    return this.api.delete<BiometricDevice>(`${this.endpoint}/${id}`);
  }
}

File: backend/ZuhooCMS/src/main/java/com/zuhoocms/modules/hrm/attendance/biometric/device/BiometricDeviceService.java
package com.zuhoocms.modules.hrm.attendance.biometric.device;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

public interface BiometricDeviceService {

    BiometricDeviceResponse create(BiometricDeviceRequest request);

    BiometricDeviceResponse getById(Long id);

    BiometricDeviceResponse getByDeviceId(String deviceId);

    Page<BiometricDeviceResponse> getAll(Pageable pageable);

    Page<BiometricDeviceResponse> getByStatus(BiometricDeviceStatus status, Pageable pageable);

    List<BiometricDeviceResponse> getOnlineDevices();

    BiometricDeviceResponse update(Long id, BiometricDeviceRequest request);

    void updateStatus(Long id, BiometricDeviceStatus status);

    void updateOnlineStatus(Long id, boolean online);

    void recordSync(Long id);

    BiometricDeviceResponse delete(Long id);
}
